// @lwc/compiler/dist/commonjs/rollup-plugins/minify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Rollup plugin applying minification to the generated bundle.
 */
function default_1({ sourcemap }) {
    // Inlining the `terser` module require to only pay the parsing and evaluation cost for needed
    // modules
    const { minify } = require('terser');
    return {
        name: 'lwc-minify',
        renderChunk(src) {
            const { code, map, error } = minify(src, {
                sourceMap: sourcemap,
                output: {
                    // Wrapping function expressions in parenthesis can be harmful for performance.
                    // Details: https://v8.dev/blog/preparser#pife
                    wrap_func_args: false,
                },
            });
            if (error) {
                throw error;
            }
            if (map) {
                return {
                    code: code,
                    map: map,
                };
            }
            else {
                return code;
            }
        },
    };
}
exports.default = default_1;
