// @lwc/template-compiler/dist/commonjs/codegen/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTemplateMetadata = exports.memorizeHandler = exports.destructuringAssignmentFromObject = exports.shouldFlatten = exports.isDynamic = exports.containsDynamicChildren = exports.isSlot = exports.isStyleSheet = exports.isTemplate = exports.objectToAST = exports.getMemberExpressionRoot = exports.kebabcaseToCamelcase = exports.identifierFromComponentName = void 0;
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const t = __importStar(require("@babel/types"));
const utils_1 = require("../shared/utils");
const ir_1 = require("../shared/ir");
const constants_1 = require("../shared/constants");
const naming_1 = require("../shared/naming");
Object.defineProperty(exports, "kebabcaseToCamelcase", { enumerable: true, get: function () { return naming_1.kebabcaseToCamelcase; } });
function identifierFromComponentName(name) {
    return t.identifier(`_${utils_1.toPropertyName(name)}`);
}
exports.identifierFromComponentName = identifierFromComponentName;
function getMemberExpressionRoot(expression) {
    let current = expression;
    while (t.isMemberExpression(current)) {
        current = current.object;
    }
    return current;
}
exports.getMemberExpressionRoot = getMemberExpressionRoot;
function objectToAST(obj, valueMapper) {
    return t.objectExpression(Object.keys(obj).map((key) => t.objectProperty(t.stringLiteral(key), valueMapper(key))));
}
exports.objectToAST = objectToAST;
/** Returns true if the passed element is a template element */
function isTemplate(element) {
    return element.tag === 'template';
}
exports.isTemplate = isTemplate;
function isStyleSheet(element) {
    return element.tag === 'style';
}
exports.isStyleSheet = isStyleSheet;
/** Returns true if the passed element is a slot element */
function isSlot(element) {
    return element.tag === 'slot';
}
exports.isSlot = isSlot;
function containsDynamicChildren(element) {
    return element.children.some((child) => ir_1.isElement(child) && isDynamic(child));
}
exports.containsDynamicChildren = containsDynamicChildren;
function isDynamic(element) {
    return !!(element.lwc && element.lwc.dynamic);
}
exports.isDynamic = isDynamic;
/**
 * Returns true if the passed element should be flattened.
 *
 * TODO [#1303]: Move this logic into the optimizing compiler. This kind of optimization should be
 * done before the actual code generation.
 */
function shouldFlatten(element) {
    return element.children.some((child) => ir_1.isElement(child) &&
        (isDynamic(child) ||
            !!child.forEach ||
            !!child.forOf ||
            (isTemplate(child) && shouldFlatten(child))));
}
exports.shouldFlatten = shouldFlatten;
function destructuringAssignmentFromObject(target, keys, type = 'const') {
    return t.variableDeclaration(type, [
        t.variableDeclarator(t.objectPattern(keys), target),
    ]);
}
exports.destructuringAssignmentFromObject = destructuringAssignmentFromObject;
function memorizeHandler(codeGen, element, componentHandler, handler) {
    // #439 - The handler can only be memorized if it is bound to component instance
    const id = getMemberExpressionRoot(componentHandler);
    const shouldMemorizeHandler = ir_1.isComponentProp(id, element);
    // Apply memorization if the handler is memorizable.
    //   $cmp.handlePress -> _m1 || ($ctx._m1 = b($cmp.handlePress))
    if (shouldMemorizeHandler) {
        const memorizedId = codeGen.getMemorizationId();
        const memorization = t.assignmentExpression('=', t.memberExpression(t.identifier(constants_1.TEMPLATE_PARAMS.CONTEXT), memorizedId), handler);
        handler = t.logicalExpression('||', memorizedId, memorization);
    }
    return handler;
}
exports.memorizeHandler = memorizeHandler;
function generateTemplateMetadata(state) {
    const metadataExpressions = [];
    // Generate the slots property on template function if slots are defined in the template:
    //      tmpl.slots = ['', 'x']
    if (state.slots.length) {
        const slotsProperty = t.memberExpression(t.identifier(constants_1.TEMPLATE_FUNCTION_NAME), t.identifier('slots'));
        const slotsArray = t.arrayExpression(state.slots.map((slot) => t.stringLiteral(slot)));
        const slotsMetadata = t.assignmentExpression('=', slotsProperty, slotsArray);
        metadataExpressions.push(t.expressionStatement(slotsMetadata));
    }
    metadataExpressions.push(...state.inlineStyle.body);
    const hasInlineStyles = state.inlineStyle.body.length;
    const stylesheetsProperty = t.memberExpression(t.identifier(constants_1.TEMPLATE_FUNCTION_NAME), t.identifier('stylesheets'));
    const stylesheetsMetadata = t.assignmentExpression('=', stylesheetsProperty, hasInlineStyles ? t.identifier('stylesheets') : t.arrayExpression());
    metadataExpressions.push(t.expressionStatement(stylesheetsMetadata));
    return metadataExpressions;
}
exports.generateTemplateMetadata = generateTemplateMetadata;
